### PR TITLE
Nix-style package management

### PR DESCRIPTION
**Pull request by jacobstanley**
Adds support for nix-style package management by creating a sandbox in a global location for each package to be installed, this means they can be shared across sandboxes using symlinks.

The idea comes from https://github.com/hvr/hackage-matrix-builder/ which does something similar.

Not sure it's ready for wider consumption just yet but I want to see if it will build on CI.

### CODE REVIEW

**Comment by charleso** [comment]
Have started testing this locally and so far so good. This is game changing!

File: src/Mafia/Init.hs [hunk 1]
```diff
@@ -2,193 +2,163 @@
...
+  packageDB     <- firstEitherT InitCabalError getPackageDB
+  packages      <- getDirectoryListing (RecursiveDepth 0) packageDB
+  packagesExist <- and <$> traverse doesFileExist packages
+  case packagesExist of
+    True  -> return PackagesOk
+    False -> return PackagesBroken
+
+------------------------------------------------------------------------
+
+data MafiaState =
```
Comment (charleso) [inline src/Mafia/Init.hs:144]: Worth having a separate module for this?

File: src/Mafia/Install.hs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  globals   <- installGlobalDependencies spkgs
+  _         <- firstEitherT InstallCabalError initSandbox
+  packageDB <- firstEitherT InstallCabalError getPackageDB
+
+  ignoreIO $ removeDirectoryRecursive packageDB
+  createDirectoryIfMissing True packageDB
+
+  env  <- getPackageEnv
+  mapM_ (link packageDB env) globals
+  Hush <- firstEitherT InstallCabalError $ cabal "sandbox" ["hc-pkg", "recache"]
```
Comment (charleso) [inline src/Mafia/Install.hs:87]: Worth adding a comment about what this does?

File: src/Mafia/Package.hs [hunk 3]
```diff
@@ -48,15 +57,38 @@ packageIdTuple (PackageId n v) =
...
     component = do
       cs <- Parse.munch1 Char.isAlphaNum
       if all Char.isDigit cs then Parse.pfail else return cs
+
+parseVersion :: Text -> Maybe Version
+parseVersion =
+  parseLongestMatch Version.parseVersion
+
+parseLongestMatch :: Parse.ReadP a -> Text -> Maybe a
+parseLongestMatch p =
```
Comment (charleso) [inline src/Mafia/Package.hs:78]: Yeah I was surprised this didn't live somewhere already?

**Comment by charleso** [comment]
Might be good to add a doc file about how the shared sandboxes work?

File: src/Mafia/Install.hs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+link :: Directory -> PackageEnv -> Package -> EitherT InstallError IO ()
+link db env p@(Package (PackageRef pid _ _) _ _) = do
+  let pcfg = packageConfigPath env p
+  unlessM (doesFileExist pcfg) (install env p)
+
+  let dest = db </> renderPackageId pid <> ".conf"
+  liftIO $ createSymbolicLink (T.unpack pcfg) (T.unpack dest)
+
+install :: PackageEnv -> Package -> EitherT InstallError IO ()
+install env p@(Package (PackageRef pid _ msrc) deps _) = do
```
Comment (charleso) [inline src/Mafia/Install.hs]: Some high level doc on this function might be good too

File: src/Mafia/Cabal/Dependencies.hs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+parseRevDeps :: Text -> Either CabalError [RevDeps]
+parseRevDeps =
+  first (CabalParseError . T.pack) .
+  traverse parseDep .
+  List.drop 1 .
+  List.dropWhile (/= "In order, the following would be installed:") .
+  T.lines
+
+parseDep :: Text -> Either String RevDeps
```
Comment (charleso) [inline src/Mafia/Cabal/Dependencies.hs]: This is probably worth some unit-level tests

**Comment by charleso** [comment]
More tests, but :100:

File: src/Mafia/Cabal/Version.hs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        Just ver -> return ver
+
+parseCabalVersion :: Text -> Maybe Version
+parseCabalVersion out =
+  case T.words out of
+    ("cabal-install" : "version" : version : _) -> parseVersion version
+    _                                           -> Nothing
+
+-- TODO not sure if maybe we should have no upper bounds here?
+-- TODO maybe mafia should just install its own cabal?
```
Comment (charleso) [inline src/Mafia/Cabal/Version.hs:43]: Yeah I guess as discussed that would be the next thing. And then mafia can...

conquer the world

File: src/Mafia/IO.hs [hunk 7]
```diff
@@ -161,8 +172,29 @@ findExecutable name = liftIO $ do
...
 getHomeDirectory :: MonadIO m => m Directory
 getHomeDirectory = T.pack `liftM` liftIO Directory.getHomeDirectory
+
+------------------------------------------------------------------------
+-- Concurrency
+
+mapConcurrentlyE :: Traversable t => (a -> EitherT x IO b) -> t a -> EitherT x IO (t b)
+mapConcurrentlyE io xs = do
+  ys <- liftIO (Async.mapConcurrently (runEitherT . io) xs)
+  hoistEither (sequence ys)
```
Comment (tranma) [inline src/Mafia/IO.hs:200]: nice

File: src/Mafia/Package.hs (hunk 3 above)
Comment (jacobstanley) [inline src/Mafia/Package.hs:78]: Yeah I'm tempted to rewrite the lot as an attoparsec parser just for consistency

File: src/Mafia/Init.hs (hunk 1 above)
Comment (jacobstanley) [inline src/Mafia/Init.hs:144]: Yeah probably, originally it was the only thing in here and was more like `InitState`

Comment (jacobstanley) [inline src/Mafia/Init.hs:144]: Hmm you know to be fair this is the only thing the init module really does (keep track of the dependency state). Maybe this should just be called something different?

**Comment by thumphries** [comment]
This looks awesome!

File: src/Mafia/Cabal/Version.hs (hunk 6 above)
Comment (markhibberd) [inline src/Mafia/Cabal/Version.hs:43]: I think installing own cabal is tremendously dangerous/not a good idea. As it stands using cabal and mafia at the same time is standard faire, and risking inconsistent versions could have particularly dire consequences. At least in the naive version of just installing its own and using it regardless of what is on path.

**Pull request merged**